### ISSUE TITLE
Full Training code

### ISSUE THREAD

**Issue by stha-prashant** [IBM/SafeLoRA#4]
Hi, could you provide the full training script for SafeLoRA?

Also I have a conceptual question about the paper.

<img width="733" height="666" alt="Image" src="https://github.com/user-attachments/assets/284e188d-8784-4e80-950e-c2b31bc5feb8" />

In this plot,  is the x-axis depth of the model? If yes, do you have any insight to the zig zag pattern of the similarity scores, is this something specific to the finetuning data or do you think this pattern may generalize to other finetuning datasets?

Thanks!!

**Comment by chiayi-hsu** [comment]
Hi,
I’m not sure whether you need the full training code or the full-parameter projection code of SafeLoRA, so I’m providing both below:

- For the full training, please refer to [LINK]( https://github.com/LLM-Tuning-Safety/LLMs-Finetuning-Safety/tree/main/llama2). When you don’t enable LoRA, it defaults to full training. 
- For full-parameter projection, please refer to [LINK](https://github.com/IBM/SafeLoRA/issues/2#issuecomment-3276056915).

For Figure 4, the x-axis represents the individual layers of the model. Here, $N$ denotes the $N$-th layer (e.g., $N=1$ is the first layer, $N=2$ is the second layer, and so on). I think this might generalize to other finetuning datasets. Recently, there has been research investigating which layers have a bigger impact on safety. You can take a look at this [paper](https://openreview.net/forum?id=kUH1yPMAn7), which could be helpful for you.
